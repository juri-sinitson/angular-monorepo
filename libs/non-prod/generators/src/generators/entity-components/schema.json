{
  "$schema": "https://json-schema.org/schema",
  "$id": "EntityComponents",
  "title": "",
  "type": "object",
  "properties": {
    "entityName": {
      "type": "string",
      "description": "The name of the component to create.",
      "x-prompt": "What component name would you like to use?"
    },
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "x-prompt": "What project name would you like to use?"
    },
    "crud": {
      "type": "boolean",
      "default": false,
      "description": "The name of the project."
    }
  },
  "required": ["entityName", "project"]
}
