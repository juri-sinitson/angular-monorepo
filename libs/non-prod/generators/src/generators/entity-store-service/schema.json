{
  "$schema": "https://json-schema.org/schema",
  "$id": "EntityStoreService",
  "title": "",
  "type": "object",
  "properties": {
    "entityName": {
      "type": "string",
      "description": "The name of the component to create.",
      "x-prompt": "What component name would you like to use?"
    },
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "x-prompt": "What project name would you like to use?"
    }
  },
  "required": ["entityName", "project"]
}
