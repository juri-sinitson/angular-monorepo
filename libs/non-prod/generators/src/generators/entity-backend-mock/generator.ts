import {
  readProjectConfiguration,
  formatFiles,
  generateFiles,
  Tree,
} from '@nx/devkit';
import * as path from 'path';
import { strings } from '@angular-devkit/core';

import { EntityBackendMockGeneratorSchema } from './schema';

export async function entityBackendMockGenerator(
  tree: Tree,
  options: EntityBackendMockGeneratorSchema
) {
  const projectConfig = readProjectConfiguration(tree, options.project);
  const projectRoot = projectConfig.root;
  const entityClassName = strings.classify(options.entityName);

  generateFiles(tree, path.join(__dirname, 'files'), projectRoot, {
    entityInterfaceName: `${entityClassName}Interface`,
    entityClassName,
    entityInUrlNamePlural: `${strings.dasherize(options.entityName)}s`,
  });
  await formatFiles(tree);
}

export default entityBackendMockGenerator;
