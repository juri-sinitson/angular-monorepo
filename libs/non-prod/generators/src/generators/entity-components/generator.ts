import {
  formatFiles,
  generateFiles,
  readProjectConfiguration,
  Tree,
} from '@nx/devkit';

import { strings } from '@angular-devkit/core';

import * as path from 'path';
import { EntityComponentsGeneratorSchema } from './schema';

export async function entityComponentsGenerator(
  tree: Tree,
  options: EntityComponentsGeneratorSchema
) {
  const projectConfig = readProjectConfiguration(tree, options.project);
  const projectRoot = projectConfig.root;

  const componentFolderName = `${strings.dasherize(options.entityName)}s`;
  const filesRoot = `${projectRoot}/src/lib/components/${componentFolderName}`;
  const entityName = strings.capitalize(options.entityName);
  generateFiles(tree, path.join(__dirname, 'files'), filesRoot, {
    className: `${strings.classify(options.entityName)}s`,
    fileName: `${strings.dasherize(options.entityName)}s`,
    entityName: strings.capitalize(options.entityName),
    entityInterfaceName: `${entityName}Interface`,
    selectorName: `${strings.dasherize(options.entityName)}s`,
    entityServiceClassName: `${strings.classify(options.entityName)}EntityStoreService`,
    entityServiceName: `${strings.camelize(options.entityName)}Service`,
  });

  await formatFiles(tree);
}

export default entityComponentsGenerator;
