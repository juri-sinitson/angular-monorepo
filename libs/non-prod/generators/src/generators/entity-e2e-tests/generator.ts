import {
  readProjectConfiguration,
  formatFiles,
  generateFiles,
  Tree,
} from '@nx/devkit';
import { strings } from '@angular-devkit/core';
import * as path from 'path';

import { EntityE2eTestsGeneratorSchema } from './schema';

export async function entityE2eTestsGenerator(
  tree: Tree,
  options: EntityE2eTestsGeneratorSchema
) {
  const projectConfig = readProjectConfiguration(tree, options.project);
  const projectRoot = projectConfig.root;
  const entitySelectorName = `${strings.dasherize(options.entityName)}`;
  const entitySelectorNamePlural = `${entitySelectorName}s`;

  const className = `${strings.classify(options.entityName)}`;  

  generateFiles(tree, path.join(__dirname, 'files'), projectRoot, {
    urlName: `${strings.camelize(options.entityName)}sUrl`,
    entitySelectorNamePlural,
    classNamePlural: `${className}s`,
  });
  await formatFiles(tree);
}

export default entityE2eTestsGenerator;
