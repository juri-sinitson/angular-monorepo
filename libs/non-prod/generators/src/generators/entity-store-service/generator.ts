import {
  readProjectConfiguration,
  formatFiles,
  generateFiles,
  Tree,
} from '@nx/devkit';
import { strings } from '@angular-devkit/core';
import * as path from 'path';

import { EntityStoreServiceGeneratorSchema } from './schema';

export async function entityStoreServiceGenerator(
  tree: Tree,
  options: EntityStoreServiceGeneratorSchema
) {
  const projectConfig = readProjectConfiguration(tree, options.project);
  const projectRoot = projectConfig.root;
  
  const filesRoot = `${projectRoot}/src/lib/stores`;
  const className = `${strings.classify(options.entityName)}`;
  const classNamePlural = `${className}s`;
  const fileName = `${strings.dasherize(options.entityName)}`;

  generateFiles(tree, path.join(__dirname, 'files'), filesRoot, {
    className,
    fileName,    
    entityName: fileName,
    classNamePlural,
    getUrlName: `getAll${classNamePlural}Url`,
    interfaceName: `${strings.classify(options.entityName)}Interface`,    
  });
  await formatFiles(tree);
}

export default entityStoreServiceGenerator;
