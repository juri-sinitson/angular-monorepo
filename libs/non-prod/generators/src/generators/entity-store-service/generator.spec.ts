import { createTreeWithEmptyWorkspace } from '@nx/devkit/testing';
import { Tree, addProjectConfiguration } from '@nx/devkit';

import { entityStoreServiceGenerator } from './generator';
import { EntityStoreServiceGeneratorSchema } from './schema';

describe('entity-store-service generator', () => {
  let tree: Tree;
  const project = 'my-lib';
  const entityName = 'MyProduct';
  const root = `libs/${project}`;

  const options: EntityStoreServiceGeneratorSchema = { 
    entityName, 
    project,
   };

  beforeEach(() => {
    tree = createTreeWithEmptyWorkspace();
  });

  it('should match snapshot ON any BY any', async () => {
    
    addProjectConfiguration(tree, project, { root });
    await entityStoreServiceGenerator(tree, options);
    
    const serviceFileContent = tree.read(`${root}/src/lib/stores/my-product-entity-store.service.ts`);    
    if(serviceFileContent) {      
      expect(serviceFileContent.toString('utf-8')).toMatchSnapshot();
    } else {
      fail('Service file not found in virtual tree!');
    }
    
    const specFileContent = tree.read(`${root}/src/lib/stores/my-product-entity-store.service.spec.ts`);
    if(specFileContent) {
      expect(specFileContent.toString('utf-8')).toMatchSnapshot();
    } else {
      fail('Spec file not found in the virtual tree!');
    }
  });
});
