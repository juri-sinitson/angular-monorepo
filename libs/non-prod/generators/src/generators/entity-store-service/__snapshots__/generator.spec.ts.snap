// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`entity-store-service generator should match snapshot ON any BY any 1`] = `
"import { Injectable } from '@angular/core';

// TODO! Adjust the project tags.
// eslint-disable-next-line @nx/enforce-module-boundaries
import { AbstractEntityStoreService } from '@angular-monorepo/shared/util-common';

// -- STEP 1: ADJUST THE URL! --
import { MyProductsUrl as entitiesUrl } from './urls';
// -- STEP 2: ADJUST THE INTERFACE PATH! --
import { MyProductInterface as EntityInterface } from '../interfaces/my-product.interface';

@Injectable({
  providedIn: 'root',
})
export class MyProductEntityStoreService extends AbstractEntityStoreService<EntityInterface> {
  protected override getDevToolsScope(): string {
    return 'MyProducts';
  }

  protected override getGetUrl(): string {
    return entitiesUrl;
  }
}
"
`;

exports[`entity-store-service generator should match snapshot ON any BY any 2`] = `
"import { TestBed } from '@angular/core/testing';
import {
  HttpTestingController,
  TestRequest,
} from '@angular/common/http/testing';

// TODO: adjust the project tags
// eslint-disable-next-line @nx/enforce-module-boundaries
import { specConfig } from '@angular-monorepo/shared/util-common-non-prod';

import { MyProductEntityStoreService as EntityStoreService } from './my-product-entity-store.service';
import { myProductsUrl as entitiesUrl } from './urls';
// TODO: adjust the project tags
// eslint-disable-next-line @nx/enforce-module-boundaries
import { MessageInterface } from '@angular-monorepo/shared/util-common';
import { MyProductInterface as EntityInterface } from '../interfaces/my-product.interface';

describe('MyProductEntityStoreService', () => {
  let service: EntityStoreService;
  let httpMock: HttpTestingController;

  const additionalEntity: EntityInterface = {
    id: '',
    code: '',
    name: 'Blue Watch',
    description: 'An extra blue watch',
    price: 72,
    category: 'Accessories',
    quantity: 61,
    inventoryStatus: 'OUTOFSTOCK',
    rating: 4,
  };

  const entitiesList: EntityInterface[] = [
    {
      id: '1000',
      code: 'f230fh0g3',
      name: 'Bamboo Watch',
      description: 'MyProduct Description',
      price: 65,
      category: 'Accessories',
      quantity: 24,
      inventoryStatus: 'INSTOCK',
      rating: 5,
    },
    {
      id: '1001',
      code: 'nvklal433',
      name: 'Black Watch',
      description: 'MyProduct Description',
      price: 72,
      category: 'Accessories',
      quantity: 61,
      inventoryStatus: 'OUTOFSTOCK',
      rating: 4,
    },
  ];

  beforeEach(() => {
    TestBed.configureTestingModule({
      providers: [specConfig.providers, OtherProductEntityStoreService],
    });
    service = TestBed.inject(OtherProductEntityStoreService);
    httpMock = TestBed.inject(HttpTestingController);
  });

  afterEach(() => {
    httpMock.verify(); // Ensure that there are no outstanding requests.
  });

  const triggerListMock = (): void => {
    const req = httpMock.expectOne(entitiesUrl);
    expect(req.request.method).toBe('GET');
    req.flush(entitiesList);
  };

  const expectError = (): void => {
    const error: MessageInterface = {
      severity: 'error',
      summary: 'Network error',
      detail: \`Http failure response for \${entitiesUrl}: 404 Not Found\`,
    };
    expect(service.messages()).toEqual([error]);
    expect(service.isError()).toBe(true);
  };

  const flushError = (req: TestRequest): void => {
    req.flush('Not Found', { status: 404, statusText: 'Not Found' });
  };

  describe('List of entities', () => {
    const pageObject = {
      triggerListMock,
      triggerListMockByEmptyList: () => {
        const req = httpMock.expectOne(entitiesUrl);
        expect(req.request.method).toBe('GET');
        req.flush([]);
      },
      triggerListError: () => {
        const req = httpMock.expectOne(entitiesUrl);
        flushError(req);
      },
      expectError,
    };

    it('should have entities ON data success BY any', () => {
      pageObject.triggerListMock();
      expect(service.entities()).toEqual(entitiesList);
      expect(service.messages()).toEqual([]);
    });

    it('should toggle loading ON data loading BY any', () => {
      expect(service.isLoading()).toBe(true);
      pageObject.triggerListMock();
      expect(service.isLoading()).toBe(false);
    });

    it('should have no entities ON data success BY no data', () => {
      pageObject.triggerListMockByEmptyList();
      expect(service.noData()).toBe(true);
    });

    it('should have error ON network error BY any', () => {
      pageObject.triggerListError();
      pageObject.expectError();
    });
  });

  describe('Create a new entity', () => {
    const pageObject = {
      triggerListMock,
      triggerEntityCreatingMock: () => {
        const req = httpMock.expectOne(entitiesUrl);
        expect(req.request.method).toBe('POST');
        expect(req.request.body).toEqual(additionalEntity);
        req.flush(additionalEntity);
      },
      triggerCreationError: () => {
        const req = httpMock.expectOne(entitiesUrl);
        expect(req.request.method).toBe('POST');
        expect(req.request.body).toEqual(additionalEntity);
        flushError(req);
      },
      expectAddedEntity: () => {
        expect(service.entities()[2]).toEqual(additionalEntity);
      },
      expectError,
    };

    it('should add entity ON data success BY any', () => {
      // Trigger the loading of the data first because we load
      // the data on init of the store
      triggerListMock();
      // Then test the adding
      service.addEntity(additionalEntity);
      pageObject.triggerEntityCreatingMock();
      pageObject.expectAddedEntity();
    });

    it('should toggle loading ON loading BY any', () => {
      // Trigger the loading of the data first because we load
      // the data on init of the store
      triggerListMock();
      // Then test the adding
      service.addEntity(additionalEntity);
      expect(service.isLoading()).toBe(true);
      pageObject.triggerEntityCreatingMock();
      expect(service.isLoading()).toBe(false);
      pageObject.expectAddedEntity();
    });

    it('should have error ON network error BY any', () => {
      // Trigger the loading of the data first because we load
      // the data on init of the store
      triggerListMock();
      // Then test the error when adding
      service.addEntity(additionalEntity);
      pageObject.triggerCreationError();
      pageObject.expectError();
    });

    it('should select uncreated entity', () => {
      pageObject.triggerListMock();
      service.selectUncreatedEntity();
      expect(service.selectedEntity()).toBe(null);
      expect(service.isNewEntityBeingEdited()).toBe(true);
    });
  });

  describe('Delete an entity', () => {
    const entityId = entitiesList[0].id;

    const pageObject = {
      triggerListMock,
      triggerDeleteMock: () => {
        const req = httpMock.expectOne(\`\${entitiesUrl}/\${entityId}\`);
        expect(req.request.method).toBe('DELETE');
        expect(req.request.url).toBe(\`\${entitiesUrl}/\${entityId}\`);
        req.flush({});
      },
      triggerDeleteError: () => {
        const req = httpMock.expectOne(\`\${entitiesUrl}/\${entityId}\`);
        expect(req.request.method).toBe('DELETE');
        flushError(req);
      },
      expectError,
      expectEntityDeleted: () => {
        expect(service.entities()).not.toContainEqual({ id: entityId });
        expect(service.entities()).toEqual([entitiesList[1]]);
      },
    };

    it('should delete entity ON delete success BY any', () => {
      // Trigger the loading of the data first because we load
      // the data on init of the store
      triggerListMock();
      // Then test the deleting
      service.deleteEntity(entityId);
      pageObject.triggerDeleteMock();
      pageObject.expectEntityDeleted();
    });

    it('should toggle loading ON loading BY any', () => {
      // Trigger the loading of the data first because we load
      // the data on init of the store
      pageObject.triggerListMock();
      // Then test the loading when deleting
      service.deleteEntity(entityId);
      expect(service.isLoading()).toBe(true);
      pageObject.triggerDeleteMock();
      expect(service.isLoading()).toBe(false);
      pageObject.expectEntityDeleted();
    });

    it('should have error ON network error BY any', () => {
      // Trigger the loading of the data first because we load
      // the data on init of the store
      pageObject.triggerListMock();
      // Then test the error when deleting
      service.deleteEntity(entityId);
      pageObject.triggerDeleteError();
      pageObject.expectError;
    });
  });

  describe('Update an entity', () => {
    const entityId = entitiesList[0].id;
    const updatedEntity: EntityInterface = {
      ...entitiesList[0],
      name: 'Updated Name',
    };

    const pageObject = {
      triggerListMock,
      triggerUpdateMock: () => {
        const req = httpMock.expectOne(\`\${entitiesUrl}/\${entityId}\`);
        expect(req.request.method).toBe('PUT');
        expect(req.request.body).toBe(updatedEntity);
        req.flush(updatedEntity);
      },
      triggerUpdateError: () => {
        const req = httpMock.expectOne(\`\${entitiesUrl}/\${entityId}\`);
        expect(req.request.method).toBe('PUT');
        flushError(req);
      },
      expectError: () => {
        const error: MessageInterface = {
          severity: 'error',
          summary: 'Network error',
          detail: \`Http failure response for \${entitiesUrl}/\${entityId}: 404 Not Found\`,
        };
        expect(service.messages()).toEqual([error]);
      },
      expectEntityUpdated: () => {
        expect(service.entities()).toContainEqual(updatedEntity);
        expect(service.entities()).not.toContainEqual(entitiesList[0]);
      },
    };

    it('should update entity ON update success BY any', () => {
      // Trigger the loading of the data first because we load
      // the data on init of the store
      pageObject.triggerListMock();
      // Then test the updating
      service.updateEntity(updatedEntity);
      pageObject.triggerUpdateMock();
      pageObject.expectEntityUpdated();
    });

    it('should toggle loading ON loading BY any', () => {
      // Trigger the loading of the data first because we load
      // the data on init of the store
      pageObject.triggerListMock();
      // Then test the loading when updating
      service.updateEntity(updatedEntity);
      expect(service.isLoading()).toBe(true);
      pageObject.triggerUpdateMock();
      expect(service.isLoading()).toBe(false);
      pageObject.expectEntityUpdated();
    });

    it('should have error ON network error BY any', () => {
      // Trigger the loading of the data first because we load
      // the data on init of the store
      pageObject.triggerListMock();
      // Then test the error when updating
      service.updateEntity(updatedEntity);
      pageObject.triggerUpdateError();
      pageObject.expectError();
    });

    // create test for selectEntityId
    it('should select entity id', () => {
      pageObject.triggerListMock();
      service.selectEntityId(entityId);
      expect(service.selectedEntity()).toEqual(entitiesList[0]);
    });

    it('should reset selected entity', () => {
      pageObject.triggerListMock();
      service.selectEntityId(entityId);
      service.resetSelectedEntity();
      expect(service.selectedEntity()).toBe(null);
      expect(service.isNewEntityBeingEdited()).toBe(false);
    });
  });
});
"
`;
